### PR TITLE
Add GLU activations

### PR DESCRIPTION
**Pull request by ruanslv**
Running ablation to check quality

### CODE REVIEW

**Review by ngoyal2707 — COMMENTED** [review]
does it work as it is without changing anything else?
cause, it will reduce the output dimension of FC1 by half (i.e 2x instead of 4x) but FC2's input dimension will still expect 4x

**Comment by stephenroller** [comment]
Yeah I agree with naman. The chunk looks weird and is maybe idiosyncratic of the source code base you got this from

**Comment by ngoyal2707** [comment]
I think functionality wise `chunk` works, as long you make sure, FC1's output projection is double than FC2's input projection.

Just readability wise, you usually dont expect activation function to preserve the dimensions

**Comment by ruanslv** [comment]
Fixed code + tried to make it prettier

**Comment by ruanslv** [comment]
Ablations worked well, SwiGLU and GeGLU provided similar results and slightly better PPL compared to GeLU. Full results in ablations sheet

File: metaseq/modules/transformer_layer.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def _linear(x, weight, bias=None):
     return F.linear(x, weight, bias)
 
 
-def _ffn(x, fc1, activation_fn, fc2, dropout_module):
+def _ffn(x, fc1, activation_fn, activation_gate, fc2, dropout_module):
```
Comment (suchenzang) [inline metaseq/modules/transformer_layer.py:29]: Nit: can we push the `activation_gate` logic into `activation_fn` instead? As in, remove the need to add an additional parameter here, along with an additional if/else conditional TransformerDecoderLayer's init method?

Reason for this request is that `TransformerDecoderLayer`'s init can get polluted really quickly with a bunch of if/else statements changing what gets constructed. Minimizing the branching factor there would help keep that logic under control.

File: metaseq/modules/transformer_layer.py [hunk 2]
```diff
@@ -46,16 +46,26 @@ def _ffn(x, fc1, activation_fn, fc2, dropout_module):
...
     else:
-        x = fc1(x)
-        x = activation_fn(x)
+        x_fc1 = fc1(x)
+        if activation_fn in utils.get_gated_activation_fns():
+            gate = activation_gate(x)
+            x = activation_fn(x_fc1, gate)
+        else:
+            x = activation_fn(x_fc1)
+        del x_fc1
```
[the remark this answers is not in the document]
Comment (ruanslv) [reply metaseq/modules/transformer_layer.py:68]: Was trying to save memory proactively, we don't need this tensor anymore after this line. Is it not needed / is this even the right way of doing it?

**Pull request closed without merging**